File: netlogo-gui/src/main/agent/World3D.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent;

import org.nlogo.core.AgentKindJ;
import org.nlogo.api.AgentException;
import org.nlogo.api.Color;
import org.nlogo.api.ImporterUser;
import org.nlogo.core.Breed;
import org.nlogo.core.Program;
import org.nlogo.api.WorldDimensionException;
import org.nlogo.core.WorldDimensions;
import org.nlogo.api.WorldDimensions3D;

import java.util.Arrays;
import java.util.Iterator;

public final strictfp class World3D
    extends World
    implements org.nlogo.api.World3D {

  Drawing3D drawing;

  public org.nlogo.api.Protractor3D protractor3D() {
    return (org.nlogo.api.Protractor3D) _protractor;
  }

  public World3D() {
    linkManager = new LinkManager3D(this);
    tieManager = new TieManager3D(this, linkManager);
    drawing = new Drawing3D(this);
    inRadiusOrCone = new InRadiusOrCone3D(this);
    _protractor = new Protractor3D(this);
    mayHavePartiallyTransparentObjects = false;
  }

  @Override
  Observer createObserver() {
    return new Observer3D(this);
  }

  @Override
  public void changeTopology(boolean xWrapping, boolean yWrapping) {
    topology = new Torus3D(this);
  }

  public void changeTopology(boolean xWrapping, boolean yWrapping, boolean zWrapping) {
    topology = new Torus3D(this);
  }

  public double shortestPathZ(double z1, double z2) {
    return ((Topology3D) topology).shortestPathZ(z1, z2);
  }

  public boolean wrappingAllowedInZ() {
    return true;
  }

  public double wrappedObserverZ(double z) {
    z = ((Topology3D) topology).wrapZ(z - followOffsetZ());

    return z;
  }

  public double followOffsetZ() {
    return ((Observer3D) _observer).followOffsetZ();
  }

  @Override
  public void diffuse4(double param, int vn) {
    throw new UnsupportedOperationException();
  }

  int _worldDepth;

  public int worldDepth() {
    return _worldDepth;
  }

  int _maxPzcor;

  public int maxPzcor() {
    return _maxPzcor;
  }

  int _minPzcor;

  public int minPzcor() {
    return _minPzcor;
  }

  Double _minPzcorBoxed;

  public Double minPzcorBoxed() {
    return _minPzcorBoxed;
  }

  Double _maxPzcorBoxed;

  public Double maxPzcorBoxed() {
    return _maxPzcorBoxed;
  }

  Double _worldDepthBoxed;

  public Double worldDepthBoxed() {
    return _worldDepthBoxed;
  }

  public double wrapZ(double z) {
    return Topology.wrap(z, _minPzcor - 0.5, _maxPzcor + 0.5);
  }

  public int roundZ(double z) {
    // floor() is slow so we don't use it
    z = ((Topology3D) topology).wrapZ(z);

    if (z > 0) {
      return (int) (z + 0.5);
    } else {
      int intPart = (int) z;
      double fractPart = intPart - z;
      return (fractPart > 0.5) ? intPart - 1 : intPart;
    }
  }

  // this procedure is the same as calling getPatchAt when the topology is a torus
  // meaning it will override the Topology's wrapping rules and
  public Patch getPatchAtWrap(double x, double y, double z) {
    int xc, yc, zc, intPart;
    double fractPart;
    x = Topology.wrap(x, _minPxcor - 0.5, _maxPxcor + 0.5);
    y = Topology.wrap(y, _minPycor - 0.5, _maxPycor + 0.5);
    z = Topology.wrap(z, _minPzcor - 0.5, _maxPzcor + 0.5);

    if (x > 0) {
      xc = (int) (x + 0.5);
    } else {
      intPart = (int) x;
      fractPart = intPart - x;
      xc = (fractPart > 0.5) ? intPart - 1 : intPart;
    }
    if (y > 0) {
      yc = (int) (y + 0.5);
    } else {
      intPart = (int) y;
      fractPart = intPart - y;
      yc = (fractPart > 0.5) ? intPart - 1 : intPart;
    }
    if (z > 0) {
      zc = (int) (z + 0.5);
    } else {
      intPart = (int) z;
      fractPart = intPart - z;
      zc = (fractPart > 0.5) ? intPart - 1 : intPart;
    }
    int patchid = ((_worldWidth * _worldHeight * (_maxPzcor - zc)) +
        (_worldWidth * (_maxPycor - yc))
        + xc - _minPxcor);


    return (Patch) _patches.toArray()[patchid];
  }

  public boolean validPatchCoordinates(int xc, int yc, int zc) {
    return
        xc >= _minPxcor &&
            xc <= _maxPxcor &&
            yc >= _minPycor &&
            yc <= _maxPycor &&
            zc >= _minPzcor &&
            zc <= _maxPzcor;
  }

  public Patch fastGetPatchAt(int xc, int yc, int zc) {
    return (Patch) _patches.toArray()[((_worldWidth * _worldHeight * (_maxPzcor - zc)) +
        (_worldWidth * (_maxPycor - yc))
        + xc - _minPxcor)];
  }

  @Override
  public Patch fastGetPatchAt(int xc, int yc) {
    return fastGetPatchAt(xc, yc, 0);
  }

  @Override
  public void createPatches(WorldDimensions dim) {
    WorldDimensions3D d = (WorldDimensions3D) dim;
    createPatches(dim.minPxcor(), dim.maxPxcor(),
                  dim.minPycor(), dim.maxPycor(),
                  d.minPzcor(), d.maxPzcor());
  }

  @Override
  public void createPatches(int minPxcor, int maxPxcor,
                            int minPycor, int maxPycor) {
    createPatches(minPxcor, maxPxcor, minPycor, maxPycor, 0, 0);
  }

  @Override
  public Program newProgram() {
    return org.nlogo.core.Program$.MODULE$.fromDialect(org.nlogo.api.NetLogoThreeDDialect$.MODULE$);
  }

  @Override
  public Program newProgram(scala.collection.Seq<String> interfaceGlobals) {
    Program emptyProgram = newProgram();
    return emptyProgram.copy(
        interfaceGlobals,
        emptyProgram.userGlobals(),
        emptyProgram.turtleVars(),
        emptyProgram.patchVars(),
        emptyProgram.linkVars(),
        emptyProgram.breeds(),
        emptyProgram.linkBreeds(),
        emptyProgram.dialect());
  }

  public void createPatches(int minPxcor, int maxPxcor,
                            int minPycor, int maxPycor,
                            int minPzcor, int maxPzcor) {
    patchScratch = null;
    patchScratch3d = null;
    _minPxcor = minPxcor;
    _maxPxcor = maxPxcor;
    _minPycor = minPycor;
    _maxPycor = maxPycor;
    _minPzcor = minPzcor;
    _maxPzcor = maxPzcor;
    _worldWidth = _maxPxcor - _minPxcor + 1;
    _worldHeight = _maxPycor - _minPycor + 1;
    _worldDepth = _maxPzcor - _minPzcor + 1;

    rootsTable = new RootsTable(_worldWidth, _worldHeight);

    _worldWidthBoxed = Double.valueOf(_worldWidth);
    _worldHeightBoxed = Double.valueOf(_worldHeight);
    _worldDepthBoxed = Double.valueOf(_worldDepth);
    _minPxcorBoxed = Double.valueOf(_minPxcor);
    _minPycorBoxed = Double.valueOf(_minPycor);
    _minPzcorBoxed = Double.valueOf(_minPzcor);
    _maxPxcorBoxed = Double.valueOf(_maxPxcor);
    _maxPycorBoxed = Double.valueOf(_maxPycor);
    _maxPzcorBoxed = Double.valueOf(_maxPzcor);


    breeds.clear();

    scala.collection.Iterator<scala.Tuple2<String, Breed>> breedIterator =
      program().breeds().iterator();

    while (breedIterator.hasNext()) {
      scala.Tuple2<String, Breed> b = breedIterator.next();
      AgentSet agentset = new TreeAgentSet(AgentKindJ.Turtle(), b._2.name());
      breeds.put(b._1.toUpperCase(), agentset);
    }

    if (_turtles != null) _turtles.clear(); // so a SimpleChangeEvent is published
    _turtles = new TreeAgentSet(AgentKindJ.Turtle(), "TURTLES");
    if (_links != null) _links.clear(); // so a SimpleChangeEvent is published
    _links = new TreeAgentSet(AgentKindJ.Link(), "LINKS");

    int x = _minPxcor;
    int y = _maxPycor;
    int z = _maxPzcor;
    Agent[] patchArray = new Agent[_worldWidth * _worldHeight * _worldDepth];
    patchColors = new int[_worldWidth * _worldHeight * _worldDepth];
    Arrays.fill(patchColors, Color.getARGBbyPremodulatedColorNumber(0.0));
    patchColorsDirty = true;

    int numVariables = program().patchesOwn().size();

    _observer.resetPerspective();

    for (int i = 0; _worldWidth * _worldHeight * _worldDepth != i; i++) {
      Patch3D patch = new Patch3D(this, i, x, y, z, numVariables);
      x++;
      if (x == (_maxPxcor + 1)) {
        x = _minPxcor;
        y--;
        if (y == (_minPycor - 1)) {
          y = _maxPycor;
          z--;
        }
      }
      patchArray[i] = patch;
    }
    _patches = new ArrayAgentSet(AgentKindJ.Patch(), patchArray, "patches");
    patchesWithLabels = 0;
    patchesAllBlack = true;
    mayHavePartiallyTransparentObjects = false;
  }

  /// export world

  @Override
  public void exportWorld(java.io.PrintWriter writer, boolean full) {
    new Exporter3D(this, writer).exportWorld(full);
  }

  @Override
  public void importWorld(org.nlogo.agent.Importer.ErrorHandler errorHandler, ImporterUser importerUser,
                          org.nlogo.agent.Importer.StringReader stringReader, java.io.BufferedReader reader)
      throws java.io.IOException {
    new Importer3D(errorHandler, this,
        importerUser, stringReader).importWorld(reader);
  }

  // used by Importer and Parser
  @Override
  public Turtle getOrCreateTurtle(long id) {
    Turtle turtle = getTurtle(id);
    if (turtle == null) {
      turtle = new Turtle3D(this, id);
    }
    return turtle;
  }

  private double[][][] patchScratch3d;

  public double[][][] getPatchScratch3d() {
    if (patchScratch3d == null) {
      patchScratch3d = new double[_worldWidth][_worldHeight][_worldDepth];
    }
    return patchScratch3d;
  }

  // these methods are primarily for behaviorspace
  // to vary the size of the world without
  // knowing quite so much about the world.
  // ev 2/20/06
  @Override
  public WorldDimensions getDimensions() {
    return new WorldDimensions3D
        (_minPxcor, _maxPxcor, _minPycor, _maxPycor, _minPzcor, _maxPzcor, 12.0, true, true, true);
  }

  @Override
  public boolean isDimensionVariable(String variableName) {
    return
        variableName.equalsIgnoreCase("MIN-PXCOR") ||
            variableName.equalsIgnoreCase("MAX-PXCOR") ||
            variableName.equalsIgnoreCase("MIN-PYCOR") ||
            variableName.equalsIgnoreCase("MAX-PYCOR") ||
            variableName.equalsIgnoreCase("MIN-PZCOR") ||
            variableName.equalsIgnoreCase("MAX-PZCOR") ||
            variableName.equalsIgnoreCase("WORLD-WIDTH") ||
            variableName.equalsIgnoreCase("WORLD-HEIGHT") ||
            variableName.equalsIgnoreCase("WORLD-DEPTH");
  }

  @Override
  public WorldDimensions setDimensionVariable(String variableName, int value, WorldDimensions d)
      throws WorldDimensionException {
    WorldDimensions3D wd = (WorldDimensions3D) d;
    if (variableName.equalsIgnoreCase("MIN-PZCOR")) {
      return new WorldDimensions3D(wd.minPxcor(), wd.maxPxcor(), wd.minPycor(), wd.maxPycor(), value, wd.maxPzcor(), 12.0, true, true, true);
    } else if (variableName.equalsIgnoreCase("MAX-PZCOR")) {
      return new WorldDimensions3D(wd.minPxcor(), wd.maxPxcor(), wd.minPycor(), wd.maxPycor(), wd.minPzcor(), value, 12.0, true, true, true);
    } else if (variableName.equalsIgnoreCase("WORLD-DEPTH")) {
      int minPzcor = growMin(wd.minPzcor(), wd.maxPzcor(), value, wd.minPzcor());
      int maxPzcor = growMax(wd.minPzcor(), wd.maxPzcor(), value, wd.maxPzcor());
      return new WorldDimensions3D(wd.minPxcor(), wd.maxPxcor(), wd.minPycor(), wd.maxPycor(), minPzcor, maxPzcor, 12.0, true, true, true);
    } else {
      WorldDimensions newWd = super.setDimensionVariable(variableName, value, d);
      return new WorldDimensions3D(newWd.minPxcor(), newWd.maxPxcor(), newWd.minPycor(), newWd.maxPycor(), wd.minPzcor(), wd.maxPzcor(), 12.0, true, true, true);
    }
  }

  @Override
  public boolean equalDimensions(WorldDimensions d) {
    return d.minPxcor() == _minPxcor &&
      d.maxPxcor() == _maxPxcor &&
      d.minPycor() == _minPycor &&
      d.maxPycor() == _maxPycor &&
      ((WorldDimensions3D) d).minPzcor() == _minPzcor &&
      ((WorldDimensions3D) d).maxPzcor() == _maxPzcor;
  }

  @Override
  public Patch getPatchAt(double x, double y)
      throws AgentException {
    return getPatchAt(x, y, 0);
  }

  public Patch3D getPatchAt(double x, double y, double z)
      throws AgentException {
    int xc = roundX(x);
    int yc = roundY(y);
    int zc = roundZ(z);

    int id = ((_worldWidth * _worldHeight * (_maxPzcor - zc)) +
        (_worldWidth * (_maxPycor - yc))
        + xc - _minPxcor);

    return (Patch3D) _patches.toArray()[id];
  }

  @Override
  public Turtle createTurtle(AgentSet breed) {
    return new Turtle3D(this, breed, ZERO, ZERO, ZERO);
  }

  // c must be in 0-13 range
  // h can be out of range
  @Override
  public Turtle createTurtle(AgentSet breed, int c, int h) {
    Turtle baby = new Turtle3D(this, breed, ZERO, ZERO, ZERO);
    baby.colorDoubleUnchecked(Double.valueOf(5 + 10 * c));
    baby.heading(h);
    return baby;
  }

  @Override
  public Object getDrawing() {
    return drawing;
  }

  // we don't ever send pixels in 3D because
  // 3D drawing is vector based. ev 5/30/06
  @Override
  public boolean sendPixels() {
    return false;
  }

  @Override
  void drawLine(double x0, double y0, double x1, double y1,
                Object color, double size, String mode) {
    drawing.drawLine(x0, y0, 0, x1, y1, 0, size, color);
  }

  void drawLine(double x0, double y0, double z0,
                double x1, double y1, double z1,
                Object color, double size) {
    drawing.drawLine(x0, y0, z0, x1, y1, z1, size, color);
  }

  @Override
  public void clearAll() {
    super.clearAll();
    drawing.clear();
  }

  @Override
  public void clearDrawing() {
    drawing.clear();
  }

  @Override
  public void stamp(Agent agent, boolean erase) {
    if (!erase) {
      drawing.stamp(agent);
    }
  }
}

File: netlogo-core/src/main/agent/World3D.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.core.{ AgentKind, Program, WorldDimensions }
import org.nlogo.api.{ AgentException, Color, ImporterUser, NetLogoThreeDDialect,
  WorldDimensionException, WorldDimensions3D }

import java.lang.{ Double => JDouble }

import java.util.Arrays

import World._

class World3D extends World
  with org.nlogo.api.World3D
  with CompilationManagement {

  val drawing: Drawing3D = new Drawing3D(this)

  override val protractor: Protractor3D = new Protractor3D(this)

  def protractor3D: org.nlogo.api.Protractor3D = protractor

  private var _patchScratch3d: Array[Array[Array[Double]]] = _

  override val linkManager = new LinkManagerImpl(this,
    { (world: World3D, src: Turtle, dest: Turtle, breed: AgentSet) =>
      val l = new Link3D(world, src, dest, breed)
      l.setId(newLinkId())
      l
    })

  protected val _links: TreeAgentSet = new TreeAgentSet(AgentKind.Link, "LINKS")

  override val tieManager: TieManager3D = new TieManager3D(_links, linkManager, protractor)
  val inRadiusOrCone = new InRadiusOrCone3D(this)

  _mayHavePartiallyTransparentObjects = false

  protected val dimensionVariableNames =
    Seq("MIN-PXCOR", "MAX-PXCOR", "MIN-PYCOR", "MAX-PYCOR", "MIN-PZCOR", "MAX-PZCOR", "WORLD-WIDTH", "WORLD-HEIGHT", "WORLD-DEPTH")

  var _worldDepth: Int = _
  var _maxPzcor: Int = _
  var _minPzcor: Int = _

  def worldDepth: Int = _worldDepth
  def minPzcor: Int = _minPzcor
  def maxPzcor: Int = _maxPzcor

  var _worldDepthBoxed: JDouble = JDouble.valueOf(_worldDepth)
  var _maxPzcorBoxed: JDouble = JDouble.valueOf(_maxPzcor)
  var _minPzcorBoxed: JDouble = JDouble.valueOf(_minPzcor)

  def worldDepthBoxed = _worldDepthBoxed
  def maxPzcorBoxed   = _maxPzcorBoxed
  def minPzcorBoxed   = _minPzcorBoxed

  _topology = new Torus3D(this)
  createPatches(_minPxcor, _maxPxcor, _minPycor, _maxPycor, _minPzcor, _maxPzcor)

  override protected def createObserver(): Observer = new Observer3D(this)

  def changeTopology(xWrapping: Boolean, yWrapping: Boolean): Unit = {
    _topology = new Torus3D(this)
  }

  def changeTopology(xWrapping: Boolean, yWrapping: Boolean, zWrapping: Boolean): Unit = {
    _topology = new Torus3D(this)
  }

  def shortestPathZ(z1: Double, z2: Double): Double =
    topology.asInstanceOf[Topology3D].shortestPathZ(z1, z2)

  val wrappingAllowedInZ = true

  def wrappedObserverZ(z: Double): Double =
    topology.asInstanceOf[Topology3D].wrapZ(z - followOffsetZ)

  def followOffsetZ: Double =
    observer.asInstanceOf[Observer3D].followOffsetZ()

  @throws(classOf[AgentException])
  @throws(classOf[PatchException])
  def diffuse(param: Double, vn: Int): Unit =
    topology.diffuse(param, vn)

  def diffuse4(param: Double, vn: Int): Unit = {
    throw new UnsupportedOperationException()
  }

  def wrapZ(z: Double): Double =
    Topology.wrap(z, _minPzcor - 0.5, _maxPzcor + 0.5)

  def wrapAndRoundZ(_z: Double): Int = {
    // floor() is slow so we don't use it
    val z = topology.asInstanceOf[Topology3D].wrapZ(_z)

    if (z > 0) {
      (z + 0.5).toInt
    } else {
      val intPart = z.toInt
      val fractPart = intPart - z
      if (fractPart > 0.5) intPart - 1 else intPart
    }
  }

  // this procedure is the same as calling getPatchAt when the topology is a torus
  // meaning it will override the Topology's wrapping rules and
  def getPatchAtWrap(x: Double, y: Double, z: Double): Patch = {
    val wrappedX = Topology.wrap(x, _minPxcor - 0.5, _maxPxcor + 0.5)
    val wrappedY = Topology.wrap(y, _minPycor - 0.5, _maxPycor + 0.5)
    val wrappedZ = Topology.wrap(z, _minPzcor - 0.5, _maxPzcor + 0.5)

    def roundBase(base: Double): Int =
      if (base > 0) {
        (base + 0.5).toInt
      } else {
        val intPart = base.toInt
        val fractPart = intPart - base
        if (fractPart > 0.5) intPart - 1 else intPart
      }

    val xc = roundBase(wrappedX)
    val yc = roundBase(wrappedY)
    val zc = roundBase(wrappedZ)
    val patchid =
      ((_worldWidth * _worldHeight * (_maxPzcor - zc)) +
        (_worldWidth * (_maxPycor - yc))
        + xc - _minPxcor)


    _patches.getByIndex(patchid).asInstanceOf[Patch]
  }

  def validPatchCoordinates(xc: Int, yc: Int, zc: Int): Boolean =
    xc >= _minPxcor && xc <= _maxPxcor &&
    yc >= _minPycor && yc <= _maxPycor &&
    zc >= _minPzcor && zc <= _maxPzcor

  def fastGetPatchAt(xc: Int, yc: Int, zc: Int): Patch =
    _patches.getByIndex(((_worldWidth * _worldHeight * (_maxPzcor - zc)) +
      (_worldWidth * (_maxPycor - yc))
      + xc - _minPxcor)).asInstanceOf[Patch]

  override def fastGetPatchAt(xc: Int, yc: Int): Patch =
    fastGetPatchAt(xc, yc, 0)

  override def createPatches(dim: WorldDimensions): Unit = {
    val d = dim.asInstanceOf[WorldDimensions3D]
    createPatches(dim.minPxcor, dim.maxPxcor,
                  dim.minPycor, dim.maxPycor,
                  d.minPzcor, d.maxPzcor)
  }

  override def createPatches(minPxcor: Int, maxPxcor: Int,
                            minPycor: Int, maxPycor: Int): Unit = {
    createPatches(minPxcor, maxPxcor, minPycor, maxPycor, 0, 0)
  }

  override def newProgram: Program =
    Program.fromDialect(NetLogoThreeDDialect)

  override def newProgram(interfaceGlobals: Seq[String]): Program =
    newProgram.copy(interfaceGlobals = interfaceGlobals)

  // 3D world copying isn't yet supported
  def copy(): World = null

  def createPatches(
    minPxcor: Int, maxPxcor: Int,
    minPycor: Int, maxPycor: Int,
    minPzcor: Int, maxPzcor: Int): Unit = {

    _patchScratch = null
    _patchScratch3d = null

    _minPxcor = minPxcor
    _maxPxcor = maxPxcor
    _minPycor = minPycor
    _maxPycor = maxPycor
    _minPzcor = minPzcor
    _maxPzcor = maxPzcor
    _worldWidth = _maxPxcor - _minPxcor + 1
    _worldHeight = _maxPycor - _minPycor + 1
    _worldDepth = _maxPzcor - _minPzcor + 1

    rootsTable = new RootsTable(_worldWidth, _worldHeight)

    _worldWidthBoxed = JDouble.valueOf(_worldWidth)
    _worldHeightBoxed = JDouble.valueOf(_worldHeight)
    _worldDepthBoxed = JDouble.valueOf(_worldDepth)
    _minPxcorBoxed = JDouble.valueOf(_minPxcor)
    _minPycorBoxed = JDouble.valueOf(_minPycor)
    _minPzcorBoxed = JDouble.valueOf(_minPzcor)
    _maxPxcorBoxed = JDouble.valueOf(_maxPxcor)
    _maxPycorBoxed = JDouble.valueOf(_maxPycor)
    _maxPzcorBoxed = JDouble.valueOf(_maxPzcor)


    breeds.clear()

    program.breeds.foreach {
      case (name, breed) =>
        val agentset = new TreeAgentSet(AgentKind.Turtle, breed.name)
        breeds.put(name.toUpperCase, agentset)
    }

    _turtles.clear() // so a SimpleChangeEvent is published
    _links.clear() // so a SimpleChangeEvent is published
    _links.directed = Directedness.Undetermined

    var x = _minPxcor
    var y = _maxPycor
    var z = _maxPzcor
    val patchArray = new Array[Agent](_worldWidth * _worldHeight * _worldDepth)
    _patchColors = new Array[Int](_worldWidth * _worldHeight * _worldDepth)
    Arrays.fill(_patchColors, Color.getARGBbyPremodulatedColorNumber(0.0))
    _patchColorsDirty = true

    val numVariables = program.patchesOwn.size

    observer.resetPerspective()

    var i = 0
    while (i < _worldWidth * _worldHeight * _worldDepth) {
      val patch = new Patch3D(this, i, x, y, z, numVariables)
      x += 1
      if (x == (_maxPxcor + 1)) {
        x = _minPxcor
        y -= 1
        if (y == (_minPycor - 1)) {
          y = _maxPycor
          z -= 1
        }
      }
      patchArray(i) = patch
      i += 1
    }
    _patches = new ArrayAgentSet(AgentKind.Patch, "patches", patchArray)
    _patchesWithLabels = 0
    _patchesAllBlack = true
    _mayHavePartiallyTransparentObjects = false
  }

  /// export world

  def exportWorld(writer: java.io.PrintWriter, full: Boolean): Unit = {
    new Exporter3D(this, writer).exportWorld(full)
  }

  @throws(classOf[java.io.IOException])
  def importWorld(errorHandler: org.nlogo.agent.ImporterJ.ErrorHandler,
    importerUser: ImporterUser,
    stringReader: org.nlogo.agent.ImporterJ.StringReader,
    reader: java.io.BufferedReader): Unit =
    new Importer3D(errorHandler, this, importerUser, stringReader).importWorld(reader)

  // used by Importer and Parser
  def getOrCreateTurtle(id: Long): Turtle = {
    val turtle = getTurtle(id)
    if (turtle == null) {
      new Turtle3D(this, id)
    } else {
      turtle
    }
  }

  def getPatchScratch3d: Array[Array[Array[Double]]] = {
    if (_patchScratch3d == null) {
      _patchScratch3d = Array.ofDim[Double](_worldWidth, _worldHeight, _worldDepth)
    }
    _patchScratch3d
  }

  // these methods are primarily for behaviorspace
  // to vary the size of the world without
  // knowing quite so much about the world.
  // ev 2/20/06
  override def getDimensions: WorldDimensions3D = {
    new WorldDimensions3D(_minPxcor, _maxPxcor, _minPycor, _maxPycor, _minPzcor, _maxPzcor, 12.0, true, true, true)
  }

  override def dimensionsAdjustedForPatchSize(patchSize: Double): WorldDimensions =
    getDimensions.copyThreeD(patchSize = patchSize)

  @throws(classOf[WorldDimensionException])
  override def setDimensionVariable(variableName: String, value: Int, d: WorldDimensions): WorldDimensions = {
    val wd = d.asInstanceOf[WorldDimensions3D]
    variableName.toUpperCase match {
      case "MIN-PZCOR" => wd.copyThreeD(minPzcor = value)
      case "MAX-PZCOR" => wd.copyThreeD(maxPzcor = value)
      case "WORLD-DEPTH" =>
        val newMin = growMin(wd.minPzcor, wd.maxPzcor, value, wd.minPzcor)
        val newMax = growMax(wd.minPzcor, wd.maxPzcor, value, wd.maxPzcor)
        wd.copyThreeD(maxPzcor = newMax, minPzcor = newMin)
      case other =>
        val newWd = super.setDimensionVariable(variableName, value, d)
        new WorldDimensions3D(
          newWd.minPxcor, newWd.maxPxcor,
          newWd.minPycor, newWd.maxPycor,
          wd.minPzcor,    wd.maxPzcor,
          12.0, true, true, true)
    }
  }

  override def equalDimensions(d: WorldDimensions): Boolean =
    d.minPxcor == _minPxcor &&
      d.maxPxcor == _maxPxcor &&
      d.minPycor == _minPycor &&
      d.maxPycor == _maxPycor &&
      d.asInstanceOf[WorldDimensions3D].minPzcor == _minPzcor &&
      d.asInstanceOf[WorldDimensions3D].maxPzcor == _maxPzcor

  @throws(classOf[AgentException])
  override def getPatchAt(x: Double, y: Double): Patch3D =
    getPatchAt(x, y, 0)

  @throws(classOf[AgentException])
  override def getPatchAt(x: Double, y: Double, z: Double): Patch3D = {
    val xc = wrapAndRoundX(x)
    val yc = wrapAndRoundY(y)
    val zc = wrapAndRoundZ(z)

    val id = ((_worldWidth * _worldHeight * (_maxPzcor - zc)) +
        (_worldWidth * (_maxPycor - yc))
        + xc - _minPxcor)

    _patches.getByIndex(id).asInstanceOf[Patch3D]
  }

  override def createTurtle(breed: AgentSet): Turtle =
    new Turtle3D(this, breed, Zero, Zero, Zero)

  // c must be in 0-13 range
  // h can be out of range
  override def createTurtle(breed: AgentSet, c: Int, h: Int): Turtle = {
    val baby = new Turtle3D(this, breed, Zero, Zero, Zero)
    baby.colorDoubleUnchecked(JDouble.valueOf(5 + 10 * c))
    baby.heading(h)
    baby
  }

  override def getDrawing: AnyRef = drawing

  // we don't ever send pixels in 3D because
  // 3D drawing is vector based. ev 5/30/06
  override def sendPixels: Boolean = false

  private[nlogo] def drawLine(
    x0: Double, y0: Double,
    x1: Double, y1: Double,
    color: AnyRef, size: Double, mode: String) {
    drawing.drawLine(x0, y0, 0, x1, y1, 0, size, color)
  }

  private[nlogo] def drawLine(
    x0: Double, y0: Double, z0: Double,
    x1: Double, y1: Double, z1: Double,
    color: AnyRef, size: Double): Unit = {
      drawing.drawLine(x0, y0, z0, x1, y1, z1, size, color)
  }

  override def clearAll(): Unit = {
    super.clearAll()
    drawing.clear()
  }

  def clearDrawing(): Unit = {
    drawing.clear()
  }

  def stamp(agent: Agent, erase: Boolean): Unit = {
    if (!erase) {
      drawing.stamp(agent)
    }
  }

  def sprout(patch: Patch, breed: AgentSet): Turtle = {
    val p3d = patch.asInstanceOf[Patch3D]
    new Turtle3D(this, breed, p3d.pxcor, p3d.pycor, p3d.pzcor)
  }
}
